Issues with vue/no-unused-components

<!--
    This template is for bug reports. If you are reporting a bug, please continue on. If you are here for another reason, please see below:

    1. To propose a new rule: http://eslint.org/docs/developer-guide/contributing/new-rules
    2. To request a change: http://eslint.org/docs/developer-guide/contributing/changes

    Note that leaving sections blank will make it difficult for us to troubleshoot and we may have to close the issue.
-->

**Tell us about your environment**

* **ESLint Version:** 5.3.0
* **eslint-plugin-vue Version:** 5.0.0-beta.3
* **Node Version:** 10.4.1

**Please show your full configuration:**

<!-- Paste your configuration below: -->
```js
module.exports = {
  root: true,
  extends: [
    'airbnb-base',
    'plugin:vue/strongly-recommended',
  ],
  parserOptions: {
    parser: 'babel-eslint',
  },
  plugins: ['vue'],
};
```

**What did you do? Please include the actual source code causing the issue.**

<!-- Paste the source code below: -->

## Issue 1

```vue
<template>
  <div id="app" class="flex flex-col h-screen">
    <Header></Header>
    <div class="flex-fill">
      <router-view></router-view>
    </div>
    <Footer></Footer>
  </div>
</template>

<script>
  import Header from 'components/Layout/Header';
  import Footer from 'components/Layout/Footer';

  export default {
    name: 'App',
    components: {
      Header,
      Footer,
    },
  };
</script>
```
Erors:

```terminal
  18:7  error  The "Header" component has been registered but not used  vue/no-unused-components
  19:7  error  The "Footer" component has been registered but not used  vue/no-unused-components
```

## Issue 2

```vue
<template>
  <div>
    <component :is="dynamicComponent"></component>
  </div>
</template>

<script>
  import Foo from 'components/Foo';
  import Bar from 'components/Bar';

  export default {
    name: 'App',
    components: {
      Foo,
      Bar,
    },
    computed: {
      dynamicComponent() {
        if (this.condition) {
          return 'Foo';
        }

        return 'Bar';
      },
    },
  };
</script>
```

Errors:

```terminal
  14:7   error  The "Foo" component has been registered but not used  vue/no-unused-components
  15:7   error  The "Bar" component has been registered but not used  vue/no-unused-components
```

**What did you expect to happen?**

Regarding issue nr. 1. 

As you can see, in the template I write components names in PascalCase because both header and footer are reserved html names. In this case I expect the plugin can detect whether component name is written in PascalCase or kebab-case.

Regarding issue nr. 2.

This is probably a very common case when someone uses dynamic component tag and uses computed property to switch between some components. In this case I think it would be great if I could somehow tell the plugin to ignore the components that are registered via special `<component />` tag.

--- a/docs/rules/no-unused-components.md
+++ b/docs/rules/no-unused-components.md
@@ -64,4 +64,50 @@ Note that components registered under other than `PascalCase` name have to be ca
 
 ## :wrench: Options
 
-Nothing.
+```json
+{
+    "vue/no-unused-components": ["error", {
+        "ignoreWhenBindingPresent": false
+    }]
+}
+```
+
+- `ignoreWhenBindingPresent` ... surpresses all error if binding has been detected in the template
+    default `false`
+
+
+:+1: Examples of **correct** code:
+
+```json
+{
+    "vue/no-unused-components": ["error", {
+        "ignoreWhenBindingPresent": true
+    }]
+}
+```
+
+```html
+<template>
+  <div>
+    <h2>Lorem ipsum</h2>
+    <component :is="computedComponent" />
+  </div>
+</template>
+
+<script>
+  import TheButton from 'components/TheButton.vue'
+  import TheSelect from 'components/TheSelect.vue'
+  import TheInput from 'components/TheInput.vue'
+
+  export default {
+    components: {
+      TheButton,
+      TheSelect,
+      TheInput,
+    },
+    computed: {
+      computedComponent() { ... }
+    }
+  }
+</script>
+```

--- a/docs/rules/no-unused-components.md
+++ b/docs/rules/no-unused-components.md
@@ -67,21 +67,21 @@ Note that components registered under other than `PascalCase` name have to be ca
 ```json
 {
     "vue/no-unused-components": ["error", {
-        "ignoreWhenBindingPresent": false
+        "ignoreWhenBindingPresent": true
     }]
 }
 ```
 
-- `ignoreWhenBindingPresent` ... surpresses all error if binding has been detected in the template
-    default `false`
+- `ignoreWhenBindingPresent` ... surpresses all errors if binding has been detected in the template
+    default `true`
 
 
-:+1: Examples of **correct** code:
+:+1: Examples of **incorrect** code:
 
 ```json
 {
     "vue/no-unused-components": ["error", {
-        "ignoreWhenBindingPresent": true
+        "ignoreWhenBindingPresent": false
     }]
 }
 ```
@@ -101,13 +101,48 @@ Note that components registered under other than `PascalCase` name have to be ca
 
   export default {
     components: {
-      TheButton,
-      TheSelect,
-      TheInput,
+      TheButton, // <- not used
+      TheSelect, // <- not used
+      TheInput, // <- not used
     },
     computed: {
       computedComponent() { ... }
     }
+  }
+</script>
+```
+
+:+1: Examples of **correct** code:
+
+```json
+{
+    "vue/no-unused-components": ["error", {
+        "ignoreWhenBindingPresent": false
+    }]
+}
+```
+
+```html
+<template>
+  <div>
+    <h2>Lorem ipsum</h2>
+    <TheButton v-if="" />
+    <TheSelect v-else-if="" />
+    <TheInput v-else="" />
+  </div>
+</template>
+
+<script>
+  import TheButton from 'components/TheButton.vue'
+  import TheSelect from 'components/TheSelect.vue'
+  import TheInput from 'components/TheInput.vue'
+
+  export default {
+    components: {
+      TheButton,
+      TheSelect,
+      TheInput,
+    },
   }
 </script>
 ```

--- a/lib/rules/no-unused-components.js
+++ b/lib/rules/no-unused-components.js
@@ -23,42 +23,50 @@ module.exports = {
       url: 'https://github.com/vuejs/eslint-plugin-vue/blob/v5.0.0-beta.3/docs/rules/no-unused-components.md'
     },
     fixable: null,
-    schema: []
+    schema: [{
+      type: 'object',
+      properties: {
+        ignoreWhenBindingPresent: {
+          type: 'boolean'
+        }
+      },
+      additionalProperties: false
+    }]
   },
 
   create (context) {
+    const options = context.options[0] || {}
     const usedComponents = []
     let registeredComponents = []
+    let ignoreReporting = false
     let templateLocation
 
     return utils.defineTemplateBodyVisitor(context, {
       VElement (node) {
-        if (!utils.isCustomComponent(node)) return
-        let usedComponentName
-
-        if (utils.hasAttribute(node, 'is')) {
-          usedComponentName = utils.findAttribute(node, 'is').value.value
-        } else if (utils.hasDirective(node, 'bind', 'is')) {
-          const directiveNode = utils.findDirective(node, 'bind', 'is')
-          if (
-            directiveNode.value.type === 'VExpressionContainer' &&
-            directiveNode.value.expression.type === 'Literal'
-          ) {
-            usedComponentName = directiveNode.value.expression.value
-          }
-        } else {
-          usedComponentName = node.rawName
+        if (utils.isHtmlElementNode(node) && !utils.isHtmlWellKnownElementName(node.rawName)) {
+          usedComponents.push(node.rawName)
         }
+      },
+      "VAttribute[directive=true][key.name='bind'][key.argument='is']" (node) {
+        if (node.value.type !== 'VExpressionContainer') return
 
-        if (usedComponentName) {
-          usedComponents.push(usedComponentName)
+        if (node.value.expression.type === 'Literal') {
+          usedComponents.push(node.value.expression.value)
+        } else if (options.ignoreWhenBindingPresent) {
+          ignoreReporting = true
         }
+      },
+      "VAttribute[directive=false][key.name='is']" (node) {
+        usedComponents.push(node.value.value)
       },
       "VElement[name='template']" (rootNode) {
         templateLocation = templateLocation || rootNode.loc.start
       },
       "VElement[name='template']:exit" (rootNode) {
-        if (rootNode.loc.start !== templateLocation) return
+        if (
+          rootNode.loc.start !== templateLocation ||
+          ignoreReporting
+        ) return
 
         registeredComponents
           .filter(({ name }) => {

--- a/lib/rules/no-unused-components.js
+++ b/lib/rules/no-unused-components.js
@@ -36,6 +36,7 @@ module.exports = {
 
   create (context) {
     const options = context.options[0] || {}
+    const ignoreWhenBindingPresent = options.ignoreWhenBindingPresent !== undefined ? options.ignoreWhenBindingPresent : true
     const usedComponents = []
     let registeredComponents = []
     let ignoreReporting = false
@@ -52,7 +53,7 @@ module.exports = {
 
         if (node.value.expression.type === 'Literal') {
           usedComponents.push(node.value.expression.value)
-        } else if (options.ignoreWhenBindingPresent) {
+        } else if (ignoreWhenBindingPresent) {
           ignoreReporting = true
         }
       },

--- a/lib/utils/index.js
+++ b/lib/utils/index.js
@@ -233,7 +233,7 @@ module.exports = {
     assert(node && node.type === 'VElement')
 
     return (
-      (this.isHtmlElementNode(node) && !this.isHtmlWellKnownElementName(node.name)) ||
+      (this.isHtmlElementNode(node) && !this.isHtmlWellKnownElementName(node.rawName)) ||
       this.hasAttribute(node, 'is') ||
       this.hasDirective(node, 'bind', 'is')
     )
@@ -280,7 +280,7 @@ module.exports = {
   isHtmlWellKnownElementName (name) {
     assert(typeof name === 'string')
 
-    return HTML_ELEMENT_NAMES.has(name.toLowerCase())
+    return HTML_ELEMENT_NAMES.has(name)
   },
 
   /**
@@ -291,7 +291,7 @@ module.exports = {
   isHtmlVoidElementName (name) {
     assert(typeof name === 'string')
 
-    return VOID_ELEMENT_NAMES.has(name.toLowerCase())
+    return VOID_ELEMENT_NAMES.has(name)
   },
 
   /**

--- a/tests/lib/rules/no-unused-components.js
+++ b/tests/lib/rules/no-unused-components.js
@@ -282,6 +282,30 @@ tester.run('no-unused-components', rule, {
         }
       </script>`
     },
+    {
+      filename: 'test.vue',
+      code: `<template>
+        <div id="app">
+          <Header></Header>
+          <div class="content">
+            <router-view></router-view>
+          </div>
+          <Footer />
+        </div>
+      </template>
+      <script>
+        import Header from 'components/Layout/Header';
+        import Footer from 'components/Layout/Footer';
+
+        export default {
+          name: 'App',
+          components: {
+            Header,
+            Footer,
+          },
+        };
+      </script>`
+    },
 
     // Ignore when `render` is used instead of temoplate
     {

--- a/tests/lib/rules/no-unused-components.js
+++ b/tests/lib/rules/no-unused-components.js
@@ -307,23 +307,6 @@ tester.run('no-unused-components', rule, {
       </script>`
     },
 
-    // Ignore when `render` is used instead of temoplate
-    {
-      filename: 'test.vue',
-      code: `
-      <script>
-        export default {
-          components: {
-            TheButton
-          },
-          render() {
-            return
-          }
-        }
-      </script>`
-    },
-
-    // Setting: ignoreWhenBindingPresent
     {
       filename: 'test.vue',
       code: `
@@ -349,6 +332,22 @@ tester.run('no-unused-components', rule, {
         }
       </script>`,
       options: [{ ignoreWhenBindingPresent: true }]
+    },
+
+    // Ignore when `render` is used instead of temoplate
+    {
+      filename: 'test.vue',
+      code: `
+      <script>
+        export default {
+          components: {
+            TheButton
+          },
+          render() {
+            return
+          }
+        }
+      </script>`
     }
   ],
   invalid: [
@@ -415,6 +414,40 @@ tester.run('no-unused-components', rule, {
       errors: [{
         message: 'The "the-button" component has been registered but not used.',
         line: 11
+      }]
+    },
+    // Setting: ignoreWhenBindingPresent
+    {
+      filename: 'test.vue',
+      code: `
+      <template>
+        <div>
+          <component :is="dynamicComponent"></component>
+        </div>
+      </template>
+      <script>
+        import Foo from 'components/Foo';
+        import Bar from 'components/Bar';
+
+        export default {
+          components: {
+            Foo,
+            Bar
+          },
+          computed: {
+            dynamicComponent() {
+              return '...'
+            }
+          }
+        }
+      </script>`,
+      options: [{ ignoreWhenBindingPresent: false }],
+      errors: [{
+        message: 'The "Foo" component has been registered but not used.',
+        line: 13
+      }, {
+        message: 'The "Bar" component has been registered but not used.',
+        line: 14
       }]
     }
   ]

--- a/tests/lib/rules/no-unused-components.js
+++ b/tests/lib/rules/no-unused-components.js
@@ -321,6 +321,34 @@ tester.run('no-unused-components', rule, {
           }
         }
       </script>`
+    },
+
+    // Setting: ignoreWhenBindingPresent
+    {
+      filename: 'test.vue',
+      code: `
+      <template>
+        <div>
+          <component :is="dynamicComponent"></component>
+        </div>
+      </template>
+      <script>
+        import Foo from 'components/Foo';
+        import Bar from 'components/Bar';
+
+        export default {
+          components: {
+            Foo,
+            Bar
+          },
+          computed: {
+            dynamicComponent() {
+              return '...'
+            }
+          }
+        }
+      </script>`,
+      options: [{ ignoreWhenBindingPresent: true }]
     }
   ],
   invalid: [

